### PR TITLE
fix: handle potential errors when creating stream

### LINKED ISSUE #1078: Errors from encodeValue_ are not caught and generate a UnhandledPromiseRejectionWarning
#### Environment details

  - OS: MacOSX 10.14.6
  - Node.js version: v10.19.0
  - npm version: 6.13.4
  - `@google-cloud/spanner` version: 5.1.0

#### Bug description
Errors thrown by [common-grpc/service/encodeValue_](https://github.com/googleapis/nodejs-spanner/blob/master/src/common-grpc/service.ts#L285) like [this one](https://github.com/googleapis/nodejs-spanner/blob/master/src/common-grpc/service.ts#L335) (passing `undefined` as value for a named parameter) are never caught by this function's caller, resulting in Unhandled Promise Rejections. **Also, for some reason it seems to crash the underlying Node.js process.**

The error thrown in `encodeValue_` goes all the way to `SessionPool.getReadSession` where it fails to be caught:
- [encodeValue_](https://github.com/googleapis/nodejs-spanner/blob/master/src/common-grpc/service.ts#L285)
- [GrpcService.encodeValue_](https://github.com/googleapis/nodejs-spanner/blob/master/src/common-grpc/service.ts#L786)
- [codec.encode](https://github.com/googleapis/nodejs-spanner/blob/master/src/codec.ts#L346)
- [Snapshot.encodeParams](https://github.com/googleapis/nodejs-spanner/blob/master/src/transaction.ts#L1034)
- [Snapshot.runStream](https://github.com/googleapis/nodejs-spanner/blob/master/src/transaction.ts#L889)
- [Database.runStream](https://github.com/googleapis/nodejs-spanner/blob/master/src/database.ts#L2109)
- [SessionPool.getReadSession](https://github.com/googleapis/nodejs-spanner/blob/master/src/session-pool.ts#L429) <-- Probably where the error should be caught

#### Steps to reproduce
There are probably different ways to reproduce the error; all you have to do is make the `encodeValue_` helper throw. One way to achieve this is to provide an `undefined` value to a named parameter in a query
  1. Run a query using named parameters, but don't pass a value to the parameter
```js
const q = {
  json: true,
  params: { },
  sql: 'SELECT * FROM tableId WHERE namedParameter = @namedParameter',
};
const [rows] = await db.run(q); // This will lead to an Unhandled Promise rejection
  ```

So any attempt to catch the error will fail
```js
const [rows] = await db.run(q).catch(() => console.log('I will never be called'));
  ```

But perhaps more unexpectedly, the code will stop running after this line:
```js
const [rows] = await db.run(q).catch(() => console.log('I will never be called'));
console.log('Unreachable code is never called ???');
  ```

I am preparing a PR which might help fix the issue; but wanted to file the issue separately in case you prefer to fix the issue differently.

Thanks for your help :)

### PR DESCRIPTION
**Pull request by olavloite**
Errors in transaction.runStream(..) were not handled by database.runStream(..), which could cause Unhandled Promise Rejections.

Fixes #1078

### CODE REVIEW

**Review by skuruppu — APPROVED** [review]

**Comment by skuruppu** [comment]
It would also be great to get feedback from @AVaksman since he was looking at this issue a few weeks ago.

**Comment by stephenplusplus** [comment]
I'm not sure exactly where the thrown error is originating from, but would it be possible to move the try/catch to what I believe is the parent method?: https://github.com/googleapis/nodejs-spanner/blob/a257e8c5924d5040e1c1dfe6570b3d3b001eb05b/src/transaction.ts

**Comment by olavloite** [comment]
> I'm not sure exactly where the thrown error is originating from, but would it be possible to move the try/catch to what I believe is the parent method?: https://github.com/googleapis/nodejs-spanner/blob/a257e8c5924d5040e1c1dfe6570b3d3b001eb05b/src/transaction.ts

I'm afraid the reference does not really point to a method, but only to a file. But assuming you mean the [runStream](https://github.com/googleapis/nodejs-spanner/blob/6e1f300ebe31027eb1f2922facb70bb0ba392b11/src/transaction.ts#L889) method in transaction.ts, then that is not a parent method of `database.runStream`. Conceptually what is going on here is the following:
1. The user application wants to execute a query or update on a database (`database.runStream(..)`)
1. The database [checks out a session from the session pool](https://github.com/googleapis/nodejs-spanner/blob/a257e8c5924d5040e1c1dfe6570b3d3b001eb05b/src/database.ts#L2167).
1. The database [starts a transaction on the session](https://github.com/googleapis/nodejs-spanner/blob/a257e8c5924d5040e1c1dfe6570b3d3b001eb05b/src/database.ts#L2173). The type of transaction is given in the options.
1. The database [executes the query / update on the transaction](https://github.com/googleapis/nodejs-spanner/blob/a257e8c5924d5040e1c1dfe6570b3d3b001eb05b/src/database.ts#L2180) that was created on the session. This is the `runStream(..)` method in transaction.ts.

The error occurs while encoding the query parameters that should be sent with the query to the server. If this fails, the query will never be sent to the server and there will never be any stream returned from the method. We need to handle that here, as we need to [return the session to the session pool](https://github.com/googleapis/nodejs-spanner/blob/a257e8c5924d5040e1c1dfe6570b3d3b001eb05b/src/database.ts#L2183).

**Comment by stephenplusplus** [comment]
Woops, sorry for the bad link and thanks for the explanation.

The method you assumed I meant is correct. That is defined on the Snapshot class, which looks to be the method being called in this PR that is now wrapped in a try/catch (`snapshot.runStream()`)? https://github.com/googleapis/nodejs-spanner/blob/a257e8c5924d5040e1c1dfe6570b3d3b001eb05b/src/database.ts#L2180

**Comment by olavloite** [comment]
@stephenplusplus What we could do is move the error handling to `snapshot.runStream` and, in case of an error, let that method return a `PartialResultStream` that will emit the error. That way, the normal error handling of the data stream can be used. WDYT?

**Comment by stephenplusplus** [comment]
I think in this case, we should make a slight adjustment to suit users who do this:

```js
const queryStream = database.runStream(...);
// later...
queryStream.on('data', () => {});
```

Users would receive an immediate, uncaught error unless they always assign an error handler right after the "runStream()" call.

Our code in the PRS file will actually wait for the stream to be reading before it attempts the network request. I think we should align our sanitization step with that cadence, something like:

```js
const {gaxOptions, json, jsonOptions, maxResumeRetries} = query;
let reqOpts;

const sanitizeRequest = () => {
  const {params, paramTypes} = Snapshot.encodeParams(query);
  const transaction: spannerClient.spanner.v1.ITransactionSelector = {};
  if (this.id) {
    transaction.id = this.id as Uint8Array;
  } else {
    transaction.singleUse = this._options;
  }
  delete query.gaxOptions;
  delete query.json;
  delete query.jsonOptions;
  delete query.maxResumeRetries;
  delete query.types;
  reqOpts = Object.assign(query, {
    session: this.session.formattedName_!,
    seqno: this._seqno++,
    transaction,
    params,
    paramTypes,
  });
}

const makeRequest = (resumeToken?: ResumeToken): Readable => {
  if (!reqOpts) {
    try {
      sanitizeRequest();
    } catch (e) {
      const errorStream = new stream.PassThrough();
      setImmediate(() => errorStream.destroy(e));
      return errorStream;
    }
  }

  return this.requestStream({
    client: 'SpannerClient',
    method: 'executeStreamingSql',
    reqOpts: Object.assign({}, reqOpts, {resumeToken}),
    gaxOpts: gaxOptions,
  });
};

return partialResultStream(makeRequest, {
  json,
  jsonOptions,
  maxResumeRetries,
});
```

**Comment by olavloite** [comment]
@stephenplusplus I updated the PR based on your suggestion. PTAL.

**Comment by stephenplusplus** [comment]
Thank you! The only thing left is to fit the tests into the appropriate `describe` in "test/transaction.ts", like:

- "should sanitize input"
- "should destroy stream with validation error"

**Comment by olavloite** [comment]
> Thank you! The only thing left is to fit the tests into the appropriate `describe` in "test/transaction.ts", like:
> 
> * "should sanitize input"
> * "should destroy stream with validation error"

Done

File: src/transaction.ts [hunk 1]
```diff
@@ -898,30 +899,42 @@ export class Snapshot extends EventEmitter {
...
-      session: this.session.formattedName_!,
-      seqno: this._seqno++,
-      transaction,
-      params,
-      paramTypes,
-    });
+    let reqOpts;
+
+    const sanitizeRequest = () => {
+      const q = query as ExecuteSqlRequest;
```
Comment (AVaksman) [inline src/transaction.ts:905]: nit: Is this step needed? It seems `query` is anyways used below line [918](https://github.com/googleapis/nodejs-spanner/pull/1208/files#diff-f5daf90433414776e2e49f79b8c93adaR918)

File: test/transaction.ts [hunk 2]
```diff
@@ -629,6 +629,40 @@ describe('Transaction', () => {
...
+      });
+
+      it('should return an error stream for invalid parameters', done => {
+        REQUEST_STREAM.resetHistory();
+
+        const fakeQuery = Object.assign({}, QUERY, {
+          params: {a: undefined},
+        });
+
+        snapshot.runStream(fakeQuery).on('error', () => done());
```
Comment (AVaksman) [inline test/transaction.ts:662]: nit: Should the error message be verified as well?
```ts
        const stream = snapshot.runStream(fakeQuery);
        stream.on('error', error => {
          assert.strictEqual(
            error.message,
            'Value of type undefined not recognized.'
          );
          done();
        });
```
Or that is excessive?

**Review by AVaksman — APPROVED** [review]

**Pull request merged**